hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java

```diff
@@ -1101,15 +1099,11 @@ public void testReportBadBlocks() throws Exception {
 
       block = DFSTestUtil.getFirstBlock(fs, filePath);
       // Test for the overloaded method reportBadBlocks
-      dataNode.reportBadBlocks(block, dataNode.getFSDataset()
-          .getFsVolumeReferences().get(0));
-      Thread.sleep(3000);
-      BlockManagerTestUtil.updateState(cluster.getNamesystem()
-          .getBlockManager());
-      // Verify the bad block has been reported to namenode
-      Assert.assertEquals(1, cluster.getNamesystem().getCorruptReplicaBlocks());
-    } finally {
-      cluster.shutdown();
+      dataNode.reportBadBlocks(block, dataNode.getFSDataset().getFsVolumeReferences().get(0));
+      DataNodeTestUtils.triggerHeartbeat(dataNode);
```

Comment:
this is why we don't need a sleep() any more, right? if so, yes, it fixes the problem deterministically

Reply:
Correct, it accounts for an internal wait to ensure namenode receives heartbeat by that time.
That's what I was referring to in previous comment, whether that internal wait needs any improvement but usually we haven't seen flakes with it. And it has multiple usages, hence we should be good IMO.

The way I am able to reproduce failure in debug mode without patch for this test, I am no longer able to reproduce test failure with the patch changes.